madara/crates/client/eth/src/harness.rs

```diff
@@ -0,0 +1,54 @@
+use alloy::node_bindings::{Anvil, AnvilInstance};
+use async_trait::async_trait;
+use tokio::sync::OnceCell;
+
+// https://etherscan.io/tx/0xcadb202495cd8adba0d9b382caff907abf755cd42633d23c4988f875f2995d81#eventlog
+// The txn we are referring to it is here ^
+pub const L1_BLOCK_NUMBER: u64 = 20395662;
+
+#[async_trait]
+pub trait AnvilInstanceInitializer: Sync + Send + 'static {
+    async fn init() -> AnvilInstance;
+}
+
+pub struct SharedAnvil<T: AnvilInstanceInitializer> {
+    cell: OnceCell<AnvilInstance>,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl<T: AnvilInstanceInitializer> SharedAnvil<T> {
+    pub const fn new() -> Self {
+        Self { cell: OnceCell::const_new(), _marker: std::marker::PhantomData }
+    }
+
+    pub async fn get_instance(&'static self) -> &'static AnvilInstance {
+        self.cell.get_or_init(T::init).await
+    }
+}
+
+impl<T: AnvilInstanceInitializer> Default for SharedAnvil<T> {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+pub struct MainnetFork;
+
+#[async_trait]
+impl AnvilInstanceInitializer for MainnetFork {
+    async fn init() -> AnvilInstance {
+        println!("Initializing Mainnet fork Anvil instance...");
+        tokio::task::spawn_blocking(create_anvil_instance).await.expect("Failed to spawn Anvil instance")
+    }
+}
+
+pub fn create_anvil_instance() -> AnvilInstance {
+    let fork_url = std::env::var("ETH_FORK_URL").expect("ETH_FORK_URL not set");
+    Anvil::new()
+        .fork(fork_url)
+        .fork_block_number(L1_BLOCK_NUMBER)
+        .port(0u16)
+        .timeout(600_000) // 10 minutes
```

Comment:
According to `anvil --help` this value applies to requests made to RPC -- I don't think 10 minutes is appropriate in that case.

```
      --timeout <timeout>
          Timeout in ms for requests sent to remote JSON-RPC server in forking mode.

          Default value 45000
```

Reply:
Actually, this seems to be a different value than what the Rust `Anvil` impl provides:

https://docs.rs/ethers-core/2.0.14/src/ethers_core/utils/anvil.rs.html#259

Reply:
Both of those links are something else. What you're looking for is: [alloy::node_bindings::Anvil](https://docs.rs/alloy-node-bindings/0.12.6/src/alloy_node_bindings/nodes/anvil.rs.html#306)

But yes it's the start up timeout, not rpc timeout. 10 minutes is overkill, but needed for when we ran sequentially. Can be relaxed down now to a lower number.

Reply:
It's the same thing, you can see here the same code snippet from `alloy`: https://docs.rs/alloy-node-bindings/0.12.6/src/alloy_node_bindings/nodes/anvil.rs.html#379

Reply:
this is alloy not ethers.rs but yes, they've appeared to reuse the same code (same team after all).